File: src/tinyquant/corpus/vector_entry.py
"""VectorEntry: entity representing a stored vector in a corpus."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from datetime import datetime

    from tinyquant._types import ConfigHash
    from tinyquant.codec.compressed_vector import CompressedVector


class VectorEntry:
    """Entity with identity by ``vector_id`` and lifecycle within the aggregate.

    Equality and hashing are by ``vector_id`` only, making entries usable
    in sets and as dict keys.
    """

    __slots__ = ("compressed", "inserted_at", "metadata", "vector_id")

    def __init__(
        self,
        vector_id: str,
        compressed: CompressedVector,
        inserted_at: datetime,
        metadata: dict[str, Any] | None = None,
    ) -> None:
        """Initialize a vector entry.

        Args:
            vector_id: Unique identity within the corpus.
            compressed: The stored compressed representation.
            inserted_at: UTC timestamp of insertion.
            metadata: Optional per-vector metadata.
        """
        self.vector_id = vector_id
        self.compressed = compressed
        self.inserted_at = inserted_at
        self.metadata = metadata

    @property
    def config_hash(self) -> ConfigHash:
        """Delegate to ``compressed.config_hash``."""
        return self.compressed.config_hash

    @property
    def dimension(self) -> int:
        """Delegate to ``compressed.dimension``."""
        return self.compressed.dimension

    @property
    def has_residual(self) -> bool:
        """Delegate to ``compressed.has_residual``."""
        return self.compressed.has_residual

    def __eq__(self, other: object) -> bool:
        """Identity equality by ``vector_id``."""
        if not isinstance(other, VectorEntry):
            return NotImplemented
        return self.vector_id == other.vector_id

    def __hash__(self) -> int:
        """Hash by ``vector_id``."""
        return hash(self.vector_id)

    def __repr__(self) -> str:
        """Return developer-friendly string representation."""
        return (
            f"VectorEntry(vector_id={self.vector_id!r}, "
            f"dimension={self.dimension}, "
            f"has_residual={self.has_residual})"
        )

File: rust/crates/tinyquant-core/src/corpus/vector_entry.rs
//! `VectorEntry` entity — one compressed vector stored in a [`Corpus`](super::Corpus).
//!
//! Identity is determined solely by [`VectorId`]; equality and hashing ignore
//! the compressed payload and metadata. This matches Python's behaviour where
//! `dict` keys are the vector identifiers.

use alloc::{collections::BTreeMap, string::String};

use crate::codec::CompressedVector;
use crate::corpus::entry_meta_value::EntryMetaValue;
use crate::types::{Timestamp, VectorId};

/// A single compressed vector together with its insertion timestamp and
/// arbitrary key-value metadata.
///
/// # Identity
///
/// `PartialEq`, `Eq`, and `Hash` are implemented **by `vector_id` only**.
/// Two entries with the same id but different payloads compare as equal.
/// This lets callers use `VectorEntry` as a set element or map key.
#[derive(Clone, Debug)]
pub struct VectorEntry {
    vector_id: VectorId,
    compressed: CompressedVector,
    inserted_at: Timestamp,
    metadata: BTreeMap<String, EntryMetaValue>,
}

impl VectorEntry {
    /// Construct a new `VectorEntry`.
    ///
    /// `inserted_at` should be a nanosecond Unix timestamp from the caller's
    /// time source.  `metadata` may be empty.
    #[must_use]
    pub const fn new(
        vector_id: VectorId,
        compressed: CompressedVector,
        inserted_at: Timestamp,
        metadata: BTreeMap<String, EntryMetaValue>,
    ) -> Self {
        Self {
            vector_id,
            compressed,
            inserted_at,
            metadata,
        }
    }

    /// The stable string identifier for this vector.
    #[must_use]
    pub const fn vector_id(&self) -> &VectorId {
        &self.vector_id
    }

    /// Reference to the compressed payload.
    #[must_use]
    pub const fn compressed(&self) -> &CompressedVector {
        &self.compressed
    }

    /// Nanosecond Unix timestamp recorded at insertion time.
    #[must_use]
    pub const fn inserted_at(&self) -> Timestamp {
        self.inserted_at
    }

    /// Read-only view of the metadata map.
    #[must_use]
    pub const fn metadata(&self) -> &BTreeMap<String, EntryMetaValue> {
        &self.metadata
    }

    /// Mutable access to the metadata map.
    pub fn metadata_mut(&mut self) -> &mut BTreeMap<String, EntryMetaValue> {
        &mut self.metadata
    }

    /// Convenience delegate: config hash of the underlying compressed vector.
    #[must_use]
    pub const fn config_hash(&self) -> &crate::types::ConfigHash {
        self.compressed.config_hash()
    }

    /// Convenience delegate: number of dimensions.
    #[must_use]
    pub const fn dimension(&self) -> u32 {
        self.compressed.dimension()
    }

    /// Convenience delegate: whether a residual buffer is present.
    #[must_use]
    pub fn has_residual(&self) -> bool {
        self.compressed.has_residual()
    }
}

impl PartialEq for VectorEntry {
    /// Equality by `vector_id` only — payload and metadata are ignored.
    fn eq(&self, other: &Self) -> bool {
        self.vector_id == other.vector_id
    }
}

impl Eq for VectorEntry {}

impl core::hash::Hash for VectorEntry {
    /// Hash by `vector_id` only — must be consistent with `PartialEq`.
    fn hash<H: core::hash::Hasher>(&self, state: &mut H) {
        self.vector_id.hash(state);
    }
}
